HorizonEditor/src/EditorLayer.cpp

```diff
@@ -0,0 +1,194 @@
+#include <pch.h>
+#include "EditorLayer.h"
+
+EditorLayer::EditorLayer(const char* name) :
+	Hzn::Layer(name),
+	m_AspectRatio((float)Hzn::App::getApp().getAppWindow().getWidth() / (float)Hzn::App::getApp().getAppWindow().getHeight()),
+	m_CameraController(Hzn::OrthographicCameraController(m_AspectRatio, 1.0f))
+{
+}
+
+void EditorLayer::onAttach()
+{
+	HZN_TRACE("Editor Layer Attached!");
+	m_CheckerboardTexture = Hzn::Texture2D::create("assets/textures/bear.png");
+	Hzn::FrameBufferProps props;
+	props.width = Hzn::App::getApp().getAppWindow().getWidth();
+	props.height = Hzn::App::getApp().getAppWindow().getHeight();
+
+	m_FrameBuffer = Hzn::FrameBuffer::create(props);
+}
+
+void EditorLayer::onDetach() {}
+
+void EditorLayer::onUpdate(Hzn::TimeStep ts)
+{
+	m_CameraController.onUpdate(ts);
+	m_FrameBuffer->bind();
+	Hzn::RenderCall::setClearColor({ 0.1f, 0.1f, 0.1f, 1.0f });
+	Hzn::RenderCall::submitClear();
+	Hzn::Renderer2D::beginScene((const Hzn::OrthographicCamera&)m_CameraController.getCamera());
+
+	for (int32_t i = 0; i < quads; ++i)
+	{
+		for (int32_t j = 0; j < quads; ++j)
+		{
+			// colored quads.
+			auto color = glm::lerp(glm::vec4(1.0f, 0.0f, 0.0f, 1.0f), glm::vec4(0.0f, 0.0f, 1.0f, 1.0f), (float)(i + j) / (2 * quads));
+			Hzn::Renderer2D::drawQuad({ i * 0.11f, j * 0.11f, 0.0f }, quadAngle, glm::vec3(0.1f), color);
+			// textured quads.
+			/*Hzn::Renderer2D::drawQuad({ i * 1.11f, j * 1.11f, 1.0f }, quadAngle, glm::vec3(1.0f), someSky);*/
+		}
+	}
+
+	Hzn::Renderer2D::endScene();
+	m_FrameBuffer->unbind();
+}
+
+bool EditorLayer::onWindowResize(Hzn::WindowResizeEvent& e)
+{
+	HZN_INFO("{0}, {1}", e.GetWidth(), e.GetHeight());
+	return false;
+}
+
+void EditorLayer::onEvent(Hzn::Event& e)
+{
+	m_CameraController.onEvent(e);
+
+	Hzn::EventDispatcher dispatcher(e);
+	dispatcher.Dispatch<Hzn::WindowResizeEvent>(std::bind(&EditorLayer::onWindowResize, this, std::placeholders::_1));
+}
+
+void EditorLayer::onRenderImgui()
+{
+	auto& window = Hzn::App::getApp().getAppWindow();
+	auto stats = Hzn::Renderer2D::getStats();
+    
+    // imgui code reference: https://github.com/ocornut/imgui/blob/docking/imgui_demo.cpp
+
+    // DOCKING CODE.
+
+    // docking options.
+    static bool p_open = true;
+    static bool opt_fullscreen = true;
+    static bool opt_padding = false;
+    static ImGuiDockNodeFlags dockspace_flags = ImGuiDockNodeFlags_None;
+
+    // We are using the ImGuiWindowFlags_NoDocking flag to make the parent window not dockable into,
+    // because it would be confusing to have two docking targets within each others.
+    ImGuiWindowFlags window_flags = ImGuiWindowFlags_MenuBar | ImGuiWindowFlags_NoDocking;
+    if (opt_fullscreen)
+    {
+        const ImGuiViewport* viewport = ImGui::GetMainViewport();
```

Comment:
This looks good

Reply:
This is the dockspace code from ImGui docking branch.